### DISCUSSION TITLE
change x text position dinamically when speed>0

### DISCUSSION THREAD

**Discussion by kessenza** [savonet/liquidsoap#2626] [General]
Using method `video.add_text` and speed > 0 (and in particular when `cycle=false`) `x` is used only as initial value but then an internal x is used to scroll text from right to the left. But when the text goes out and we, for instance, want to change it (to show a different info text) then the internal x stays out of the screen and there is no way to bring it back to the right

**Comment by smimram** [comment]
For  this one, let's first discuss this here. I fear that if I add this the code will become clumsy and moreover, we will never be able to cover all possible cases of needs with `video.add_text`.

*However*, I have good news: you can achieve what you want in Liquidsoap, by programming your own advanced `x` parameter. The `x` function is called at every frame, you can then simply return the value of a reference, updated in a way which suits your needs. You can draw inspiration from the implementation of `video.add_text` itself: https://github.com/savonet/liquidsoap/blob/main/src/libs/video.liq#L301

Typically for a scrolling you would do something like:
```python
fps = video.frame.rate()
r = ref(0)
speed = 100 # scrolling speed

def x()
  x := !x + speed / fps
  !x
end
video.add_text(x, "text", main_source)
```
and you can easily add more logic in order to detect when the boundary is reached etc.

**Comment by kessenza** [comment]
Just to be sure (because I am little weak on ls scripting language). Do you mean ? :

```
fps = video.frame.rate()
r = ref(0)
speed = 100 # scrolling speed

def x()
  r := !r + speed / fps
  !r
end
```

Yes this solves all the possible cases together wtih #2621 
The only little thing is that in this way (in particular when) cycling on the opposite way (` r := !r - speed / fps)`, it's not easy to detect when text goes out of the screen because you have to compute `string.length(text) * pixel_by_char`

Anyway it works. Thanks!

**Reply by smimram** [reply]
Yes (sorry I did not test my code).

I have converted this issue into a discussion so that we can freely chat :)

**Comment by smimram** [comment]
For the issue with text size, my advice would be to start from [the code of `video.add_text`](https://github.com/savonet/liquidsoap/blob/main/src/libs/video.liq#L301)

The idea is that you use `video.text` to generate the text and then add it on top of your main source. The advantage is that you easily have acccess to the size of the video now with `video.dimensions`.

So, it would be something like this (once again, not tested):
```python
fps = video.frame.rate()

# The source containing text
text = video.text("my nice text")
# Add width and height methods
text = video.dimensions(video.crop(text))

# Translate the text
r = ref(0)
speed = 100 # scrolling speed
def x()
  r := !r - speed / fps
  if !r < - text.width() then ... end # cycle
  !r
end
text = video.translate(x=x, text)

# Finally add the text to your main source
s = add([main_source, text])
```
Thats it!

**Reply by kessenza** [reply]
Thanks for your time explaining how this could be achieved.
I agree and this fits exactly my needs.
Just a couple of questions. Not sure what translate do ? Documentation doesn't help sorry.

Why do you need to crop video before get dimensions? It seems that this works in the same way and gives same result without cropping

```
# The source containing text
text = video.text("my nice text")
# Add width and height methods
text = video.dimensions(text)
```

**Reply by smimram** [reply]
If you don't crop, you will get the default dimensions of an image instead of the actual dimensions of the text. Moreover, cropping is almost free because of our internal representation of images (there is no actual suppression of pixels, etc.)

**Reply by kessenza** [reply]
ok clear. And what does translate do? I read the book and search all over the documentation but couldn't figure what "it translates the video" mean. Thanks

**Reply by smimram** [reply]
`video.text` writes the upper-right corner. `video.translate` moves it to the right `x` / `y` position. (see [here](https://en.wikipedia.org/wiki/Translation_(geometry)) for what a translation is, although I am not sure this is very helpful)

**Comment by kessenza** [comment]
Thanks for your reply. Now it's all perfectly clear.
